Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/person/ui/details/items/PersonDetailsQrCard.kt

```diff
@@ -68,6 +70,48 @@ class PersonDetailsQrCard(parent: ViewGroup) :
                     )
                 }
             }
+
+            when (certificate.getState()) {
+                is CwaCovidCertificate.State.ExpiringSoon -> {
+                    expirationStatusIcon.visibility = View.VISIBLE
+                    expirationStatusIcon.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_av_timer))
+                    expirationStatusText.visibility = View.VISIBLE
+                    expirationStatusText.text = context.getString(
+                        R.string.certificate_qr_expiration,
+                        curItem.certificate.headerExpiresAt.toShortDayFormat(),
+                        curItem.certificate.headerExpiresAt.toShortTimeFormat()
+                    )
+                }
+
+                is CwaCovidCertificate.State.Expired -> {
+                    expirationStatusIcon.visibility = View.VISIBLE
+                    expirationStatusIcon.setImageDrawable(
+                        ContextCompat.getDrawable(
+                            context,
+                            R.drawable.ic_error_outline
+                        )
+                    )
+                    expirationStatusText.visibility = View.VISIBLE
+                    expirationStatusText.text = context.getText(R.string.certificate_qr_expired)
+                    qrSubtitle.visibility = View.GONE
+                }
+
+                is CwaCovidCertificate.State.Invalid -> {
+                    expirationStatusIcon.visibility = View.VISIBLE
+                    expirationStatusIcon.setImageDrawable(
+                        ContextCompat.getDrawable(
+                            context,
+                            R.drawable.ic_error_outline
+                        )
+                    )
+                    expirationStatusText.visibility = View.VISIBLE
+                    expirationStatusText.text = context.getText(R.string.certificate_qr_invalid_signature)
+                    qrSubtitle.visibility = View.GONE
+                }
+
+                else -> {
+                }
```

Comment:
should we handle `Valid` case also to hide validity  views and show `qrSubtitle`?
In normal cases state should not change from `Invalid`, `Expired`, `ExpiringSoon` to `Valid` again , but the device time is a factor here , some (users/ testers) would act as time travellers which could lead to having some views visible for old state

Reply:
Haven't thought about this but you're right, I'll handle it as well.